usb/backend/libusb0.py

```diff
@@ -638,7 +638,7 @@ def is_kernel_driver_active(self, dev_handle, intf):
                         length))
             return cast(name, c_char_p).value != b'usbfs'
         except USBError as err:
-            if err.backend_error_code == -ENODATA:
+            if err.backend_error_code == -errno.errorcode[errno.ENODATA]:
```

Comment:
How can `errno.ENODATA` still be valid if `from errno import ENODATA` fails?

According to the documentation (and to a quick look at the CPython code), that failure implies that ENODATA is not defined for your platform.

Also, on your platform, what do `errno.ENODATA` and `errno.errorcode[errno.ENODATA]` equal to?

Reply:
True!!! Good catch! Sorry and thank you :-)

Python 3.7:
```
(venv37zephyr) python
Python 3.7.10 (default, Apr  8 2021, 01:15:14)
[Clang 10.0.1 (git@github.com:llvm/llvm-project.git llvmorg-10.0.1-0-gef32c611a on freebsd12
Type "help", "copyright", "credits" or "license" for more information.
>>> import errno
>>> errno.ENODATA
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: module 'errno' has no attribute 'ENODATA'
>>> errno.errorcode[errno.ENODATA]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: module 'errno' has no attribute 'ENODATA'
```

Python 3.8:
```
hexagon% source venv38bsd/bin/activate
(venv38bsd) hexagon% python
Python 3.8.9 (default, Apr 15 2021, 01:10:27)
[Clang 10.0.1 (git@github.com:llvm/llvm-project.git llvmorg-10.0.1-0-gef32c611a on freebsd12
Type "help", "copyright", "credits" or "license" for more information.
>>> import errno
>>> errno.ENODATA
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: module 'errno' has no attribute 'ENODATA'
>>> errno.errorcode[errno.ENODATA]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: module 'errno' has no attribute 'ENODATA'
```

This looks OS dependent value :-(
```
(venv37zephyr) uname -a
FreeBSD hexagon 12.2-RELEASE-p6 FreeBSD 12.2-RELEASE-p6 GENERIC  amd64
(venv37zephyr) less /usr/include/errno.h
(venv37zephyr) grep ENODATA /usr/include/errno.h
(venv37zephyr) ls -al /usr/include/errno.h
lrwxr-xr-x  1 root  wheel  11 23 lis  2018 /usr/include/errno.h -> sys/errno.h
```

This problem occurs here https://github.com/pyocd/pyOCD/blob/e8797c61083d320a665cd5832ee346d8e074b3ae/pyocd/probe/pydapaccess/interface/pyusb_backend.py#L99 when pyOCD wants to detach kernel driver that may interfere with its operations. Another problem here is that standard user is not supposed to attach/detach kernel drivers. I did put try/except around that and when kernel is not detached things work as expected anyway. 

Now I can see that I have considered fix working because different error occurs and only `except (NotImplementedError, usb.core.USBError) as e:` is handled:
```
Type "help", "copyright", "credits" or "license" for more information.
>>> import errno
>>> errno.ENODATA
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: module 'errno' has no attribute 'ENODATA'
>>> from errno import ENODATA
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: cannot import name 'ENODATA' from 'errno' (unknown location)
```

How can we fix that? How can we fix that on BSD that already has no value defined?

Looks again like no change is necessary on `pyusb` side but the client application..? :-)

Reply:
* Okay so after some analysis I can see that this part of `pyusb` may need some improvement.
* Current `pyusb/backend/libusb0.py` code is tied to Linux, which is one problem. `libusb1.py` does not use `ENODATA` at all.
* Another problem is that import is not handled in a way it could raise an exception, so I cannot handle it anyhow from the application level, everything just crashes. 
* FreeBSD implementation differs from Linux implementation and somehow we are passing the `if not hasattr(_lib, 'usb_get_driver_np')` check so it means the function is implemented. Can you please hint us @hselasky on what it returns - is it `usbfs` when in use or `ENODATA` alternative when not used by kernel module?
* Physical interface is running DAPLink firmware - it uses HID (CMSIS-DAP Debug that we are interested in this part), VCP (Serial Console Port to microcontroller), and UMS (drag-n-drop firmware flashing).

Reply:
Actually, this is not the only problem: our implementation assumed `usb_get_driver_np()` would only be available on Linux, and makes no sense for FreeBSD's libusb0.

FreeBSD's `usb_get_driver_np()` [appears](https://svnweb.freebsd.org/base/head/lib/libusb/usb.h?revision=326219&view=markup) to simply wrap `libusb20_dev_get_iface_desc()`, which in turn is [documented](https://www.unix.com/man-page/freebsd/3/libusb20_dev_get_iface_desc/) to do:

> `libusb20_dev_get_iface_desc()` retrieves the kernel interface description for the given USB `iface_index`.  The format of the USB interface description is: `"drivername<unit>: <description>"` The description string is always zero terminated.  A zero length string is written in case no driver is attached to the given interface.  The USB device given by `pdev` must be opened before this function will succeed.  This function returns zero on success else a `LIBUSB20_ERROR` value is returned.

Additionally, the `usbfs` driver is a special case on Linux: while it's a kernel driver, it's the one used by libusb.  As that's not a driver that needs to be detached, and following the libusb1 API, our function returns `False` in that case.  A similar consideration may apply to FreeBSD...

Reply:
I have tried to enforce libusb1 backend (https://github.com/pyocd/pyOCD/blob/e8797c61083d320a665cd5832ee346d8e074b3ae/pyocd/probe/pydapaccess/interface/pyusb_backend.py#L73) but that does not seem to work and this could save us some work. Note that LibUSB-1.0 is supported on FreeBSD.
```
 72         # Get device handle
 73         if "BSD" in platform.system():
 74             import usb.backend.libusb1 as libusb1
 75             b = libusb1.get_backend()
 76             if b is None:
 77                 print("LIBUSB1 BACKEND UNAVAILABLE!")
 78                 exit(-1)
 79             dev = usb.core.find(custom_match=FindDap(self.serial_number),
 80                 backend=b)
 81         else:
 82             dev = usb.core.find(custom_match=FindDap(self.serial_number))
```
And the result is:
```
(venv37zephyr) pudb3 -m pyocd flash ../zephyr.hex
LIBUSB1 BACKEND UNAVAILABLE!
```
or without printf we can see that libusb0 is selected anyway returning us to the starting point:
```
(venv37zephyr) pudb3 -m pyocd flash ../zephyr.hex
0001962:CRITICAL:__main__:uncaught exception: cannot import name 'ENODATA' from 'errno' (unknown location)
Traceback (most recent call last):
  File "/mnt/stuff/x/cederom-pyocd.git/pyocd/__main__.py", line 150, in run
    status = cmd.invoke()
  File "/mnt/stuff/x/cederom-pyocd.git/pyocd/subcommands/flash_cmd.py", line 86, in invoke
    with session:
  File "/mnt/stuff/x/cederom-pyocd.git/pyocd/core/session.py", line 343, in __enter__
    self.open()
  File "/mnt/stuff/x/cederom-pyocd.git/pyocd/core/session.py", line 457, in open
    self._probe.open()
  File "/mnt/stuff/x/cederom-pyocd.git/pyocd/probe/cmsis_dap_probe.py", line 170, in open
    self._link.open()
  File "/mnt/stuff/x/cederom-pyocd.git/pyocd/utility/concurrency.py", line 28, in _locking
    return func(self, *args, **kwargs)
  File "/mnt/stuff/x/cederom-pyocd.git/pyocd/probe/pydapaccess/dap_access_cmsis_dap.py", line 635, in open
    self._interface.open()
  File "/mnt/stuff/x/cederom-pyocd.git/pyocd/probe/pydapaccess/interface/pyusb_backend.py", line 108, in open
    if dev.is_kernel_driver_active(interface_number):
  File "/home/x/usr/local/venv37zephyr/lib/python3.7/site-packages/pyusb-1.1.1-py3.7.egg/usb/core.py", line 1100, in is_kernel_driver_active
    interface)
  File "/home/x/usr/local/venv37zephyr/lib/python3.7/site-packages/pyusb-1.1.1-py3.7.egg/usb/backend/libusb0.py", line 624, in is_kernel_driver_active
    from errno import ENODATA
ImportError: cannot import name 'ENODATA' from 'errno' (unknown location)
```
Help :-)

Reply:
What is the `.so` name for FreeBSD's libusb1?

Reply:
I took a look at libusb-0.1.12 and there is no `libusb_kernel_driver_active()` yet, probably this is why `usb_get_driver_np` is used? But we have that new function on FreeBSD. Why detection of `libusb1` backend fails?
```
% grep libusb_kernel_driver_active /usr/include/*usb*
/usr/include/libusb.h:int       libusb_kernel_driver_active(libusb_device_handle * devh, int interface);
```

Reply:
It fails because PyUSB (using `ctypes.util.find_library()`) can't find the appropriate `.so`.

What is the `.so` name for FreeBSD's libusb1?

Reply:
> What is the `.so` name for FreeBSD's libusb1?

```
(venv37zephyr) ls /usr/lib/*usb*
/usr/lib/libusb_p.a     /usr/lib/libusb.so      /usr/lib/libusbhid_p.a  /usr/lib/libusbhid.so
/usr/lib/libusb.a       /usr/lib/libusb.so.3    /usr/lib/libusbhid.a    /usr/lib/libusbhid.so.4
```

Reply:
> FreeBSD's `usb_get_driver_np()` [appears](https://svnweb.freebsd.org/base/head/lib/libusb/usb.h?revision=326219&view=markup) to simply wrap `libusb20_dev_get_iface_desc()`, which in turn is [documented](https://www.unix.com/man-page/freebsd/3/libusb20_dev_get_iface_desc/) (..)

Yup :-) There are two variants of this function:
1. `usb_get_driver_np` contained in `libusb01.c` calls `libusb20_dev_get_iface_desc()` and truncates the result (reference: https://cgit.freebsd.org/src/tree/lib/libusb/libusb01.c#n981)
2. `libusb_get_driver_np` contained in `libusb10.c` simply wraps the `libusb_get_driver()` (reference: https://cgit.freebsd.org/src/tree/lib/libusb/libusb10.c#n871).

> Additionally, the `usbfs` driver is a special case on Linux: while it's a kernel driver, it's the one used by libusb. As that's not a driver that needs to be detached, and following the libusb1 API, our function returns `False` in that case. A similar consideration may apply to FreeBSD...

Yes, except this Linux `ENODATA` return code which on FreeBSD seems to be simply `0` or `-1`. Maybe a simple if wrap can be applied here to distinguish the import + response?

If we solve this `libusb0` issue it would be also nice to solve `libusb1` issue and use new library as the default backend.. can we do that without `libusb_get_parent()`?

Reply:
Btw. when I `python setup.py install` I get package stripped out of source code and I am unable to debug modified `pyusb` anymore - how can I solve that? :-)

Reply:
> Yup :-) There are two variants of this function:

I wouldn't call them variants.  The first is part of the cross-platform libusb0 API, the second is something specific to FreeBSD's libusb1.

> Yes, except this Linux ENODATA return code which on FreeBSD seems to be simply 0 or -1.

From the documentation of `libusb20_dev_get_iface_desc()`, I think the return value can be computed solely from the returned string (taking into account possible usbfs-like behavior).  And, unlike on Linux, any errors can simply be forwarded to the caller (as `USBError`).

Can you prepare and test a patch for this?

> Maybe a simple if wrap can be applied here to distinguish the import + response?

Not sure what you meant there.

> If we solve this libusb0 issue it would be also nice to solve libusb1 issue and use new library as the default backend

Sure, we should fix that too.

> can we do that without `libusb_get_parent()`?

I think so.

By the way, I missed your first mention to `libusb_get_parent()` because I had already answered to a previous version of that comment.  Try to avoid edits that add relevant/brand new information, especially in fairly active issues.

Reply:
> Btw. when I `python setup.py install` I get package stripped out of source code and I am unable to debug modified `pyusb` anymore - how can I solve that? :-)

Do you mean with [`pdb`](https://docs.python.org/3/library/pdb.html)?

And have you installed this "modified" PyUSB or are you having problems importing it instead of the installed version?

Reply:
> > Yes, except this Linux ENODATA return code which on FreeBSD seems to be simply 0 or -1.
> From the documentation of `libusb20_dev_get_iface_desc()`, I think the return value can be computed solely from the returned string (taking into account possible usbfs-like behavior). And, unlike on Linux, any errors can simply be forwarded to the caller (as `USBError`).

Here is the source:
```
usb_get_driver_np(usb_dev_handle * dev, int interface, char *name, int namelen)
{
	struct libusb20_device *pdev;
	char *ptr;
	int err;

	pdev = (void *)dev;

	if (pdev == NULL)
		return (-1);
	if (namelen < 1)
		return (-1);
	if (namelen > 255)
		namelen = 255;

	err = libusb20_dev_get_iface_desc(pdev, interface, name, namelen);
	if (err != 0)
		return (-1);

	/* we only want the driver name */
	ptr = strstr(name, ":");
	if (ptr != NULL)
		*ptr = 0;

	return (0);
}
```

> Can you prepare and test a patch for this?

Will do :-) Just need to take some sleep :-)

> > If we solve this libusb0 issue it would be also nice to solve libusb1 issue and use new library as the default backend
> Sure, we should fix that too.

:-)


> > can we do that without `libusb_get_parent()`?
> I think so.

:-)

> By the way, I missed your first mention to `libusb_get_parent()` because I had already answered to a previous version of that comment. Try to avoid edits that add relevant/brand new information, especially in fairly active issues.

ACK :-)

Reply:
> > Btw. when I `python setup.py install` I get package stripped out of source code and I am unable to debug modified `pyusb` anymore - how can I solve that? :-)
> 
> Do you mean with [`pdb`](https://docs.python.org/3/library/pdb.html)?

Well I am using ncurses based `pudb3` (https://pypi.org/project/pudb/) but this is the same + text gui for faster navigation (like TUI+GDB / CGDB). Instead `import pdb; pdb.set_trace()` I am calling `import pudb; pudb.set_trace()`.

> And have you installed this "modified" PyUSB or are you having problems importing it instead of the installed version?

I am working in Python VirtualEnv, usually I just install packages with `pip install pyusb` but for testing I am removing them with `pip uninstall pyusb` and then I install modified version for testing with `python setup.py install`. I need to install `pyusb` as this is required package for `pyocd`. If package is no dependency for others I can simply call it with `pudb3 -m pyocd` directly from the source code repository :-)

Reply:
`python setup.py develop` is the answer in place of `python setup.py install` - that will install development version of module with debug symbols included :-)

Reply:
> > Can you prepare and test a patch for this?

Done :-) Please verify @jonasmalacofilho :-)

Reply:
Thanks!

Reply:
Also, I really need to stop using GitHub's "request changes" option when submitting a review.  While its description is more or less accurate,

> feedback that must be addressed before merging.

I think that's only visible to me, and I'm obviously not _requesting_ anything.  Instead, I'm _asking_ for you to make these changes.

: )

Reply:
no worries, whatever button, we press it :-P